### PR TITLE
[SD 3.5 Dreambooth LoRA] support configurable training block & layers

### PR DESCRIPTION
**Pull request by linoytsaban**
This PR adds

-  `--lora_layers` and `--lora_blocks` to dreambooth lora training script to allow targeting specific blocks & layers.
-  This is generally proven to be useful, and specifically for SD3.5 initial thoughts on the matter brought up here - [Stable Diffusion 3.5 Large Fine-tuning Tutorial](https://stabilityai.notion.site/Stable-Diffusion-3-5-Large-Fine-tuning-Tutorial-11a61cdcd1968027a15bdbd7c40be8c6)
- some initial results to demonstrate the differences: 

Group 3-14

### CODE REVIEW

**Comment by linoytsaban** [comment]
cc @bghira - maybe you explored and have some insights/thoughts to share too 🤗

File: examples/dreambooth/README_sd3.md [hunk 1]
```diff
@@ -147,6 +147,41 @@ accelerate launch train_dreambooth_lora_sd3.py \
...
 
+### Targeting Specific Blocks & Layers
+As image generation models get bigger & more powerful, more fine-tuners come to find that training only part of the 
+transformer blocks (sometimes as little as two) can be enough to get great results. 
+In some cases, it can be even better to maintain some of the blocks/layers frozen.
+
+For **SD3.5-Large** specifically, you may find this information useful (taken from: [Stable Diffusion 3.5 Large Fine-tuning Tutorial](https://stabilityai.notion.site/Stable-Diffusion-3-5-Large-Fine-tuning-Tutorial-11a61cdcd1968027a15bdbd7c40be8c6#12461cdcd19680788a23c650dab26b93):
+> [!NOTE]
+> A commonly believed heuristic that we verified once again during the construction of the SD3.5 family of models is that later/higher layers (i.e. `30 - 37`)* impact tertiary details more heavily. Conversely, earlier layers (i.e. `12 - 24` )* influence the overall composition/primary form more. 
+> So, freezing other layers/targeting specific layers is a viable approach.
```
Comment (sayakpaul) [inline examples/dreambooth/README_sd3.md:158]: Might want to expand this with something like:

When targeting for X, we should generally target Y layers. 

WDYT?

File: examples/dreambooth/README_sd3.md [hunk 2]
```diff
@@ -147,6 +147,41 @@ accelerate launch train_dreambooth_lora_sd3.py \
...
+> `*`These suggested layers are speculative and not 100% guaranteed. The tips here are more or less a general idea for next steps.
+> **Photorealism**
+> In preliminary testing, we observed that freezing the last few layers of the architecture significantly improved model training when using a photorealistic dataset, preventing detail degradation introduced by small dataset from happening.
+> **Anatomy preservation**
+> To dampen any possible degradation of anatomy, training only the attention layers and **not** the adaptive linear layers could help. For reference, below is one of the transformer blocks.
```
Comment (sayakpaul) [inline examples/dreambooth/README_sd3.md:163]: Makes total sense to me!

File: examples/dreambooth/README_sd3.md [hunk 3]
```diff
@@ -147,6 +147,41 @@ accelerate launch train_dreambooth_lora_sd3.py \
...
+
+
+we've added `--lora_layers` and `--lora_blocks` to make LoRA training modules configurable.
```
Comment (sayakpaul) [inline examples/dreambooth/README_sd3.md:166]:
```suggestion
We've added `--lora_layers` and `--lora_blocks` to make LoRA training modules configurable. 
```

File: examples/dreambooth/README_sd3.md [hunk 4]
```diff
@@ -147,6 +147,41 @@ accelerate launch train_dreambooth_lora_sd3.py \
...
+> In preliminary testing, we observed that freezing the last few layers of the architecture significantly improved model training when using a photorealistic dataset, preventing detail degradation introduced by small dataset from happening.
+> **Anatomy preservation**
+> To dampen any possible degradation of anatomy, training only the attention layers and **not** the adaptive linear layers could help. For reference, below is one of the transformer blocks.
+
+
+we've added `--lora_layers` and `--lora_blocks` to make LoRA training modules configurable. 
+- with `--lora_blocks` you can specify the block numbers for training. E.g. passing - 
+```bash
+--lora_blocks="12,13,14,15,16,17,18,19,20,21,22,23,24,30,31,32,33,34,35,36,37"
+```
```
Comment (sayakpaul) [inline examples/dreambooth/README_sd3.md:170]:
```suggestion

```diff
+ --lora_blocks="12,13,14,15,16,17,18,19,20,21,22,23,24,30,31,32,33,34,35,36,37"
```

```

Comment (sayakpaul) [reply examples/dreambooth/README_sd3.md:170]: Verify the change when you apply it. I think there might be some backquote rendering problems.

File: examples/dreambooth/README_sd3.md [hunk 5]
```diff
@@ -147,6 +147,41 @@ accelerate launch train_dreambooth_lora_sd3.py \
...
+will trigger LoRA training of transformer blocks 12-24 and 30-37. By default, all blocks are trained. 
+- with `--lora_layers` you can specify the types of layers you wish to train. 
+By default, the trained layers are -  
+`"attn.add_k_proj","attn.add_q_proj","attn.add_v_proj", "attn.to_add_out","attn.to_k","attn.to_out.0","attn.to_q","attn.to_v"`
+If you wish to have a leaner LoRA / train more blocks over layers you could pass -
```
Comment (sayakpaul) [inline examples/dreambooth/README_sd3.md:175]: Leaner LoRA targetting what aspect? From what I understand, this heuristic is for targeting a specific quality, right?

File: examples/dreambooth/README_sd3.md [hunk 6]
```diff
@@ -147,6 +147,41 @@ accelerate launch train_dreambooth_lora_sd3.py \
...
+--lora_blocks="12,13,14,15,16,17,18,19,20,21,22,23,24,30,31,32,33,34,35,36,37"
+```
+will trigger LoRA training of transformer blocks 12-24 and 30-37. By default, all blocks are trained. 
+- with `--lora_layers` you can specify the types of layers you wish to train. 
+By default, the trained layers are -  
+`"attn.add_k_proj","attn.add_q_proj","attn.add_v_proj", "attn.to_add_out","attn.to_k","attn.to_out.0","attn.to_q","attn.to_v"`
+If you wish to have a leaner LoRA / train more blocks over layers you could pass - 
+```bash
+--lora_layers="attn.to_k,attn.to_q,attn.to_v,attn.to_out.0"
+```
```
Comment (sayakpaul) [inline examples/dreambooth/README_sd3.md:178]: Same. Let's make it a "diff".

File: examples/dreambooth/README_sd3.md [hunk 7]
```diff
@@ -147,6 +147,41 @@ accelerate launch train_dreambooth_lora_sd3.py \
...
+This will reduce LoRA size by roughly 50% for the same rank compared to the default. 
+However, if you're after compact LoRAs, it's our impression that maintaining the default setting for `--lora_layers` and
+freezing some of the early & blocks is usually better. 
+
+The following configuration
```
Comment (sayakpaul) [inline examples/dreambooth/README_sd3.md:183]: Incomplete?

File: examples/dreambooth/test_dreambooth_lora_sd3.py [hunk 8]
```diff
@@ -136,6 +136,42 @@ def test_dreambooth_lora_latent_caching(self):
...
+            lora_state_dict = safetensors.torch.load_file(os.path.join(tmpdir, "pytorch_lora_weights.safetensors"))
+            is_lora = all("lora" in k for k in lora_state_dict.keys())
+            self.assertTrue(is_lora)
+
+            # when not training the text encoder, all the parameters in the state dict should start
+            # with `"transformer"` in their names.
+            # In this test, only params of transformer block 0 should be in the state dict
+            starts_with_transformer = all(
+                key.startswith("transformer.transformer_blocks.0") for key in lora_state_dict.keys()
+            )
```
Comment (sayakpaul) [inline examples/dreambooth/test_dreambooth_lora_sd3.py:172]: Super!

File: examples/dreambooth/train_dreambooth_lora_sd3.py [hunk 9]
```diff
@@ -571,6 +571,25 @@ def parse_args(input_args=None):
         "--adam_weight_decay_text_encoder", type=float, default=1e-03, help="Weight decay to use for text_encoder"
     )
 
+    parser.add_argument(
+        "--lora_layers",
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sd3.py:575]: Let's add a test for this too?

Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_sd3.py:575]: Let's add a test for this too?

**Review by sayakpaul — COMMENTED** [review]
Thank you! Left a couple of comments. LMK if they make sense or are unclear.

File: examples/dreambooth/README_sd3.md (hunk 5 above)
Comment (linoytsaban) [reply examples/dreambooth/README_sd3.md:175]: Indeed the feature was added to allow experimentation with what layers produce the best quality, but since the default (once the pr is merged) will be 
`attn.add_k_proj attn.add_q_proj attn.add_v_proj attn.to_add_out attn.to_k attn.to_out.0 attn.to_q attn.to_v`
which makes every trained block chunkier than the previous default, I wanted to also give as an example the previous setting we had which is 
`--lora_layers attn.to_k attn.to_q attn.to_v attn.to_out.0`
that will result in a smaller lora. 
But if it's confusing/unclear I can remove that

File: examples/dreambooth/README_sd3.md [hunk 10]
```diff
@@ -147,6 +147,40 @@ accelerate launch train_dreambooth_lora_sd3.py \
...
+```diff
+--lora_blocks "12,13,14,15,16,17,18,19,20,21,22,23,24,30,31,32,33,34,35,36,37"
+```
+will trigger LoRA training of transformer blocks 12-24 and 30-37. By default, all blocks are trained. 
+- with `--lora_layers` you can specify the types of layers you wish to train. 
+By default, the trained layers are -  
+`attn.add_k_proj,attn.add_q_proj,attn.add_v_proj,attn.to_add_out,attn.to_k,attn.to_out.0,attn.to_q,attn.to_v`
+If you wish to have a leaner LoRA / train more blocks over layers you could pass - 
+```diff
+--lora_layers attn.to_k,attn.to_q,attn.to_v,attn.to_out.0
```
Comment (sayakpaul) [inline examples/dreambooth/README_sd3.md:177]: Are we deleting this block or we're suggesting to add it? In case of addition, it should be like so `+ --lora_layers attn.to_k,attn.to_q,attn.to_v,attn.to_out.0` under the diff block.

File: examples/dreambooth/test_dreambooth_lora_sd3.py [hunk 11]
```diff
@@ -136,6 +136,76 @@ def test_dreambooth_lora_latent_caching(self):
...
+            self.assertTrue(os.path.isfile(os.path.join(tmpdir, "pytorch_lora_weights.safetensors")))
+
+            # make sure the state_dict has the correct naming in the parameters.
+            lora_state_dict = safetensors.torch.load_file(os.path.join(tmpdir, "pytorch_lora_weights.safetensors"))
+            is_lora = all("lora" in k for k in lora_state_dict.keys())
+            self.assertTrue(is_lora)
+
+            # when not training the text encoder, all the parameters in the state dict should start
+            # with `"transformer"` in their names.
+            # In this test, only params of transformer block 0 should be in the state dict
```
Comment (sayakpaul) [inline examples/dreambooth/test_dreambooth_lora_sd3.py:205]: Revisit the comment?

File: examples/dreambooth/test_dreambooth_lora_sd3.py [hunk 12]
```diff
@@ -136,6 +136,76 @@ def test_dreambooth_lora_latent_caching(self):
...
+            test_args = f"""
+                {self.script_path}
+                --pretrained_model_name_or_path {self.pretrained_model_name_or_path}
+                --instance_data_dir {self.instance_data_dir}
+                --instance_prompt {self.instance_prompt}
+                --resolution 64
+                --train_batch_size 1
+                --gradient_accumulation_steps 1
+                --max_train_steps 2
+                --lora_blocks 0
```
Comment (sayakpaul) [inline examples/dreambooth/test_dreambooth_lora_sd3.py:150]: I would make the value we're passing here a constant so that it's clear from the test as to what we're varying. Does that make sense?

File: examples/dreambooth/test_dreambooth_lora_sd3.py [hunk 13]
```diff
@@ -136,6 +136,76 @@ def test_dreambooth_lora_latent_caching(self):
...
+            test_args = f"""
+                {self.script_path}
+                --pretrained_model_name_or_path {self.pretrained_model_name_or_path}
+                --instance_data_dir {self.instance_data_dir}
+                --instance_prompt {self.instance_prompt}
+                --resolution 64
+                --train_batch_size 1
+                --gradient_accumulation_steps 1
+                --max_train_steps 2
+                --lora_layers attn.to_k
```
Comment (sayakpaul) [inline examples/dreambooth/test_dreambooth_lora_sd3.py:186]: Same as https://github.com/huggingface/diffusers/pull/9762/files#r1819028559.

**Review by sayakpaul — APPROVED** [review]
Excellent work here. I love it, especially the tests! I left only minor comments, LMK if they make sense.
